File: p079m/word_search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function (board, word) {
  const visited = Array.from(board, (r) => Array.from(r, (c) => false));
  const [rows, cols] = [board.length, board[0].length];

  // Check if word starting from [iw] available starting board[ib, jb],
  // board, word are accessed under closure
  const recExist = function (i, j, c) {
    if (c === word.length) return true;
    if (i < 0 || j < 0 || i >= rows || j >= cols) return false;
    if (visited[i][j] || board[i][j] !== word[c]) return false;

    visited[i][j] = true;
    // DFS search for the next char with backtracking
    const found =
      recExist(i - 1, j, c + 1) || // up
      recExist(i, j - 1, c + 1) || // left
      recExist(i + 1, j, c + 1) || // down
      recExist(i, j + 1, c + 1); // right
    visited[i][j] = false;
    return found;
  };

  // Initiate the search
  for (let i = 0; i < board.length; ++i) {
    for (let j = 0; j < board[i].length; ++j) {
      if (recExist(i, j, 0)) return true;
    }
  }
  return false;
};
// TEST
[
  [
    [
      ['A', 'B', 'C', 'E'],
      ['S', 'F', 'C', 'S'],
      ['A', 'D', 'E', 'E'],
    ],
    'ABCCED',
    true,
  ],
  [
    [
      ['A', 'B', 'C', 'E'],
      ['S', 'F', 'C', 'S'],
      ['A', 'D', 'E', 'E'],
    ],
    'SEE',
    true,
  ],
  [
    [
      ['A', 'B', 'C', 'E'],
      ['S', 'F', 'C', 'S'],
      ['A', 'D', 'E', 'E'],
    ],
    'ABCB',
    false,
  ],
].forEach(([board, word, expected]) => {
  const actual = exist(board, word);
  console.log('Word', word, 'exist in the board ->', actual);
  console.assert(actual === expected);
});

File: p079m/word_search.py
class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])
        dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]

        def helper(i: int, j: int, c: int) -> bool:
            if c == len(word):
                return True
            if i < 0 or i >= rows or j < 0 or j >= cols:
                return False
            if board[i][j] != word[c]:
                return False
            mem = board[i][j]
            board[i][j] = "#"
            for di, dj in dirs:
                if helper(i + di, j + dj, c + 1):
                    return True
            board[i][j] = mem
            return False

        for i in range(rows):
            for j in range(cols):
                if helper(i, j, 0):
                    return True
        return False


# TESTS
for board, word, expected in [
    (
        [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]],
        "ABCCED",
        True,
    ),
    (
        [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]],
        "SEE",
        True,
    ),
    (
        [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]],
        "ABCB",
        False,
    ),
    (
        [
            ["A", "A", "A", "A", "A", "A"],
            ["A", "A", "A", "A", "A", "A"],
            ["A", "A", "A", "A", "A", "A"],
            ["A", "A", "A", "A", "A", "A"],
            ["A", "A", "A", "A", "A", "A"],
            ["A", "A", "A", "A", "A", "A"],
        ],
        "AAAAAAAAAAAABAA",
        False,
    ),
]:
    sol = Solution()
    actual = sol.exist(board, word)
    print("Word", word, "exists in the board ->", actual)
    assert actual == expected
